packages/shopping-cart/src/use-shopping-cart.ts

```diff
@@ -1,11 +1,49 @@
-import { useContext } from 'react';
+import debugFactory from 'debug';
+import { useContext, useState, useEffect, useRef, useMemo } from 'react';
 import ShoppingCartContext from './shopping-cart-context';
-import type { ShoppingCartManager } from './types';
+import ShoppingCartOptionsContext from './shopping-cart-options-context';
+import useRefetchOnFocus from './use-refetch-on-focus';
+import type { UseShoppingCart } from './types';
 
-export default function useShoppingCart(): ShoppingCartManager {
-	const manager = useContext( ShoppingCartContext );
-	if ( ! manager ) {
+const debug = debugFactory( 'shopping-cart:use-shopping-cart' );
+
+export default function useShoppingCart( cartKey?: string | undefined ): UseShoppingCart {
+	const managerClient = useContext( ShoppingCartContext );
+	if ( ! managerClient ) {
 		throw new Error( 'useShoppingCart must be used inside a ShoppingCartProvider' );
 	}
-	return manager;
+
+	const { defaultCartKey } = useContext( ShoppingCartOptionsContext ) ?? {};
+	const finalCartKey = cartKey ?? defaultCartKey;
+	debug( `getting cart manager for cartKey ${ finalCartKey }` );
+	const manager = managerClient.forCartKey( finalCartKey );
+
+	// Re-render when the cart changes
+	const isMounted = useRef< boolean >( true );
+	useEffect( () => {
+		isMounted.current = true;
+		return () => {
+			isMounted.current = false;
+		};
+	}, [] );
+	const [ cartState, setCartState ] = useState( manager.getState() );
+	useEffect( () => {
+		if ( finalCartKey ) {
+			debug( 'subscribing to cartKey', finalCartKey );
+			return managerClient.forCartKey( finalCartKey ).subscribe( () => {
+				debug( 'cart manager changed; re-rendering' );
+				isMounted.current && setCartState( manager.getState() );
+			} );
+		}
+	}, [ managerClient, finalCartKey, manager ] );
+
+	useRefetchOnFocus( finalCartKey );
+
+	return useMemo(
+		() => ( {
+			...cartState,
+			...manager.actions,
+		} ),
+		[ cartState, manager ]
+	);
```

Comment:
> Further, the manager could have a method that produces the UseShoppingCart value directly, so that we don't need to construct it in the hook.

This was how I had it originally, but it felt awkward because we had to regenerate the whole thing every time the state changed, even though the actions didn't need to be regenerated. I refactored it to split up state and actions in the manager, and then combined them in the hook to keep the existing API. I suppose I could just move this combination into a separate method in the manager. 🤔

Reply:
> I suppose I could just move this combination into a separate method in the manager.

I did this but I'm not sure that I like the result. It's more elegant to have the hook value's creation be entirely done inside the manager object, but that means that the manager object has to have that method as part of its public API. If it's part of the public API, then someone eventually will probably try to use it directly. If someone tries to use it directly, they're likely to create bugs because of the time-based nature of the state as described in my example above.

The hook needs to work with and cache the manager's data in a specific way (combine time-based state with static actions whenever the time-based state changes) that's really a private implementation detail that should not be exposed to consumers. I'm working to see if there's a sensible way to extract the caching and calculation into a (non-public) stand-alone function but that might end up taking a lot more code than the hook would use to do the same thing.

Reply:
I think I resolved this by creating a `ShoppingCartHookManager` that can deal with creating and caching the hook state.